### PR TITLE
feat: fused argmin and argmax to avoid unnecessary memory allocation

### PR DESCRIPTION
**Pull request by eddyxu**
Implement fused argmin / argmax that takes an iterator instead of an array, to avoid memory allocation.

### CODE REVIEW

File: rust/lance-linalg/src/kmeans.rs [hunk 1]
```diff
@@ -446,37 +447,44 @@ impl KMeans {
...
-                                    MetricType::Cosine => vector.cosine(other),
-                                    MetricType::Dot => vector.dot(other),
-                                };
-                                if dist < min {
-                                    min = dist;
-                                    min_idx = idx;
-                                }
-                            }
-                            (min_idx as u32, min)
+                            argmin_value(centroids_array.chunks_exact(dimension).enumerate().map(
```
Comment (eddyxu) [inline rust/lance-linalg/src/kmeans.rs:477]: about 30-40% speed up for fast cosine route.

File: rust/lance-linalg/src/kernels.rs [hunk 2]
```diff
@@ -47,24 +58,54 @@ where
...
+///
+pub fn argmin<T: Num + PartialOrd + Copy>(iter: impl Iterator<Item = T>) -> Option<u32>
 where
-    T::Native: PartialOrd + Bounded,
+    T: Bounded,
 {
+    argmin_value(iter).map(|(idx, _)| idx)
+}
+
+pub fn argmin_value<T: Num + Bounded + PartialOrd + Copy>(
```
Comment (chebbyChefNEQ) [inline rust/lance-linalg/src/kernels.rs:72]: this function seems to be the same as `argmin_value_opt `?

Comment (eddyxu) [reply rust/lance-linalg/src/kernels.rs:72]: One takes `Iterator<Item=Option<T>>` the other takes `Iterator<Item=T>`.   Arrow's primitive array `Float32Array::iter()` is a `Iterator<Option<T>>`.

Comment (eddyxu) [reply rust/lance-linalg/src/kernels.rs:72]: Can probably just let `argmin_val` to call `argmin_val_opt` with little performance penalty tho.

Comment (chebbyChefNEQ) [reply rust/lance-linalg/src/kernels.rs:72]: oh, i was expected the opt is in the return argument. ok this makes sense.

File: rust/lance-linalg/src/kmeans.rs [hunk 3]
```diff
@@ -446,37 +447,44 @@ impl KMeans {
...
                             let vector = &array[idx * dimension..(idx + 1) * dimension];
-                            let mut min = std::f32::MAX;
-                            let mut min_idx = 0;
-                            for (idx, other) in centroids_array.chunks_exact(dimension).enumerate()
-                            {
-                                // We've found about 40% performance improvement by using static dispatch instead
-                                // of dynamic dispatch.
-                                //
-                                // NOTE: Please make sure run benchmark when changing the following code.
-                                // `RUSTFLAGS="-C target-cpu=native" cargo bench --bench ivf_pq`
```
Comment (chebbyChefNEQ) [inline rust/lance-linalg/src/kmeans.rs:468]: should we keep this note here?

+ have we tested branching cost? i.e. if we have a generic `KMeans<DistanceType=...>` vs match statement?

File: rust/lance-linalg/src/kmeans.rs [hunk 4]
```diff
@@ -446,37 +447,44 @@ impl KMeans {
         let n = data.len() / self.dimension;
         let metric_type = self.metric_type;
         const CHUNK_SIZE: usize = 1024;
+
+        // Normalized centroids for fast cosine.
```
Comment (chebbyChefNEQ) [inline rust/lance-linalg/src/kmeans.rs:451]: uber nit: clarify this allows us to skip one of the norm calculation in `A.B/(|A||B|)`

**Review by chebbyChefNEQ — APPROVED** [review]
just nits

File: rust/lance-linalg/src/kmeans.rs (hunk 3 above)
Comment (wjones127) [reply rust/lance-linalg/src/kmeans.rs:468]: +1 on this. For critical pieces of code, I bet it's worth not having the `match` in the loop.

Comment (eddyxu) [reply rust/lance-linalg/src/kmeans.rs:468]: > have we tested branching cost? i.e. if we have a generic KMeans<DistanceType=...> vs match statement?

Can do in follow up PRs.

This one meant to make #1312 smaller / reviewable.

File: rust/lance-linalg/src/kmeans.rs (hunk 4 above)
Comment (eddyxu) [reply rust/lance-linalg/src/kmeans.rs:451]: Added

File: rust/lance-linalg/src/kmeans.rs (hunk 3 above)
Comment (eddyxu) [reply rust/lance-linalg/src/kmeans.rs:468]: added the comment back

**Pull request merged**